src/System.IO.Pipes/src/System/IO/Pipes/PipeCompletionSource.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Security;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Runtime.InteropServices;
+
+namespace System.IO.Pipes
+{
+    internal abstract unsafe class PipeCompletionSource<TResult> : TaskCompletionSource<TResult>
+    {
+        private const int NoResult = 0;
+        private const int ResultSuccess = 1;
+        private const int ResultError = 2;
+        private const int RegisteringCancellation = 4;
+        private const int CompletedCallback = 8;
+
+        private readonly CancellationToken _cancellationToken;
+        private readonly ThreadPoolBoundHandle _threadPoolBinding;
+
+        private CancellationTokenRegistration _cancellationRegistration;
+        private NativeOverlapped* _overlapped;
+        private int _state;
+
+#if DEBUG
+        private bool _cancellationHasBeenRegistered;
+#endif
+
+        // Using RunContinuationsAsynchronously for compat reasons (old API used ThreadPool.QueueUserWorkItem for continuations)
+        protected PipeCompletionSource(ThreadPoolBoundHandle handle, CancellationToken cancellationToken, object pinData)
+            : base(TaskCreationOptions.RunContinuationsAsynchronously)
+        {
+            Debug.Assert(handle != null, "handle is null");
+
+            _threadPoolBinding = handle;
+            _cancellationToken = cancellationToken;
+            _state = NoResult;
+
+            _overlapped = _threadPoolBinding.AllocateNativeOverlapped((errorCode, numBytes, pOverlapped) =>
+            {
+                var completionSource = (PipeCompletionSource<TResult>)ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped);
+                Debug.Assert(completionSource.Overlapped == pOverlapped);
+
+                completionSource.AsyncCallback(errorCode, numBytes);
+            }, this, pinData);
+        }
+
+        internal NativeOverlapped* Overlapped
```

Comment:
Is this Overlapped property accessed anywhere except for this base type now?  If so, could we just get rid of the property and have all accesses be to the _overlapped field?